### ISSUE TITLE
Report a warning, if only symlinks are being backed up

### ISSUE THREAD

**Issue by viric** [restic/restic#542]
Yesterday I thought I was backing up a directory...Yesterday I thought I was backing up a directory...let's say, ~/mydoc, and I forgot that ~/mydoc is just a symlink to ~/doc/own

So I was backing up a symlink... It happened for several days in a row so I don't have the backups I expected.

I wish restic had said me something like: "you are backing up a symlink to a directory, not a directory".

Definitely it correctly reported "0 directories 1 file" backed up, but it took me some days to notice.

**Comment by cfcs** [comment]
This sounds like a nice warning. Perhaps a `--follow-symlinks` would also be nice?

**Comment by fd0** [comment]
Let's track implementing the warning in this issue, and leave following symlinks for another one.

**Comment by alphapapa** [comment]
Just ran into this myself.  One of my most important directories is a symlink in my homedir that points to the real directory deeper in my homedir tree.  Thankfully I caught it while testing restic, rather than after I started depending on it.

I'm not sure what the "correct" behavior would be; I can imagine some users would want to backup and restore symlinks rather than what they point to.  But I also imagine that some users will expect the files/directories pointed to by symlinks to be backed up, and that they will be in for a nasty surprise someday when they need to run a restore.

So I realize it's not an easy call, but I'm not sure I feel safe relying on restic if I have to keep this limitation in mind every time I create a symlink.

**Comment by cfcs** [comment]
Remember to ignore the warning condition if the target of the symlink is indeed included in the backup set.
- You should pay attention to how `exclude` patterns may differ for the two directories (what happens if I exclude `~/.config` and include `~/allmyimportantthings` which happens to point to a directoy in `~/.config` ?)

**Comment by alphapapa** [comment]
> You should pay attention to how exclude patterns may differ for the two directories (what happens if I exclude ~/.config and include ~/allmyimportantthings which happens to point to a directoy in ~/.config ?)

Yes, what does happen in that case?

**Comment by fd0** [comment]
Thanks for all the input, but I'm not sure if having an option to follow symlinks is a good idea after all. It is very easy to create endless loops, like this:
```
$ tree .
.
├── dir1
│   └── x -> ../dir2
└── dir2
    └── y -> ../dir1
```

I've looked at what `tar` does, it has an option to dereference symlinks:
```
$ tar cv --dereference -f /tmp/test.tar .
```

Called this way it will never terminate but happily keep archiving and endless loop of `x/y/x/y/x/y`. Is that even a good idea?

**Comment by pvgoran** [comment]
I'm not a fan of dereferencing symlinks either, but I think it's worth mentioning that you can create an endless loop situation even without it. It can be set up using `mount --bind`.

I remember one of standard filesystem-traversing programs (find? rsync? something other?) detected such a thing, issued a warning and skipped the (rest of the) infinite traversal loop. Probably restic needs to do have this logic as well, regardless of dereferencing symlinks or not.

**Comment by alphapapa** [comment]
> I'm of the impression that having an option to dereference symlinks causes a lot more problems and corner cases that need to be handled in the code, unfortunately.

You may be right.  May I propose an option to issue warnings for each symlink that is backed up?  Probably should be off by default, but for cautious users, they could enable it and check the log after running a backup to make sure they didn't accidentally backup any symlinks instead of actual data.

**Comment by cfcs** [comment]
@alphapapa I think warnings for symlinks that point *out of the backed-up data* is a very good idea. Perhaps also warnings for non-relative symlinks since that may cause problems when restoring to a different directory.

**Comment by chneukirchen** [comment]
I just ran into this too, and it's even worse as `restic backup path/to/symlink` is stored as a symlink (plausible), but also `restic backup path/to/symlink/` (note the trailing slash) is stored as a symlink, contrary to almost all other Unix tools!

**Comment by pvgoran** [comment]
While this kind of warnings can be quite useful in certain situations, I doubt they should be enabled by default. Both types of symlinks (out of backed-up data, and absolute) are perfectly fine when doing system-wide backups, when the user knows exactly where he has data and what he wants to backup.

**Comment by fd0** [comment]
@chneukirchen Can you be a bit more specific which tools use a trailing slash to indicate that a symlink should be followed? It think it is not so universal as you may expect: At least GNU `tar`, of which I'm taking a lot of inspiration since it has been around forever, saves a symlink in both cases: `tar cvf /tmp/x.tar path/to/symlink` and `tar cvf /tmp/x.tar path/to/symlink/`. I can imagine that this is a special case implemented by e.g. `rsync`, which distinguishes a directory and its contents with a trailing slash.

**Comment by viric** [comment]
"ls" for example.

ls -ld mysymlink

ls -ld mysymlink/

On Sat, Aug 26, 2017 at 03:19:43AM -0700, Alexander Neumann wrote:
> @chneukirchen Can you be a bit more specific which tools use a trailing slash to indicate that a symlink should be followed? It think it is not so universal as you may expect: At least GNU `tar`, of which I'm taking a lot of inspiration since it has been around forever, saves a symlink in both cases: `tar cvf /tmp/x.tar path/to/symlink` and `tar cvf /tmp/x.tar path/to/symlink/`. I can imagine that this is a special case implemented by e.g. `rsync`, which distinguishes a directory and its contents with a trailing slash.
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/restic/restic/issues/542#issuecomment-325111736

-- 
(Escriu-me xifrat si saps PGP / Write ciphered if you know PGP)
PGP key 7CBD1DA5 - https://emailselfdefense.fsf.org/

**Comment by chneukirchen** [comment]
I was thinking about ls and find.  But yes, tar behaves as you say.  So perhaps add a `-H` ;)

**Comment by fd0** [comment]
I've spent some time thinking about this and came up with a different idea: There are two underlying problems that users have which they are trying to solve with symlinks:
 * Make the directory structure match some desired structure
 * Include the target of the symlinks in the backup

First, I think it may be a good idea to print a warning message when a symlinks is saved and the target of the symlink exists and is not included in the backup. This way, users will know that some data referenced in the snapshot is not available upon restore.

The second of the problems outlined above can be resolved much more elegantly: Add an option to restic to include the targets of symlinks into the backup. This solves the data accessibility nicely (all referenced data is available after restore) and, much more important, the restored directory structure matches the structure of the original data.

This would be a mid-term goal because it requires internal changes in the way restic works.

What do you think?

**Comment by pvgoran** [comment]
> Add an option to restic to include the targets of symlinks into the backup.
Do you mean adding both the symlinks and their targets?

**Comment by fd0** [comment]
Yes, that's what I suggested.

**Comment by pvgoran** [comment]
How would the symlinked files/directories represented in the snapshot, then? What if the command line has just directory A which has multiple symlinks whose targets are all over the filesystem (in directories B, C, D, ...)? Will all these directories (B, C, D, ...) appear as additional top-level directories in the snapshot? This would be messy.

Also, where would the symlinked files/directories be restored? I don't see a reliable way to restore the symlinks and symlinked files/directories so that the former would still point to the latter.

**Comment by matthijskooijman** [comment]
A related issue happens on Windows when using VSS snapshots. There is a Powershell script in #340 which creates a VSS snapshot to backup from, and to give the backup a stable name, a symlink to it is created. However, when you want to backup the *entire* snapshot, you pass the name of the symlink to restic, which does not follow it.

One additional thought is that some tools have an option to follow no symlinks (usually the default), all symlinks, or only symlinks specified on the commandline directly. The latter option is fairly safe (does not follow arbitrary symlinks), covers the VSS usecase as well as some others and makes sense - if you specified a link on the commandline, you probably weren't trying to get just the link backed up.

**Comment by zaucks** [comment]
> I've spent some time thinking about this and came up with a different idea: There are two underlying problems that users have which they are trying to solve with symlinks:
> 
> * Make the directory structure match some desired structure
> * Include the target of the symlinks in the backup

Hi,
  I think you could add the same 'links' options as rsync (https://download.samba.org/pub/rsync/rsync.html), at least '--copy-unsafe-links' which does what is discussed here :
> tells rsync to copy the referent of symbolic links that point outside the copied tree

> It is very easy to create endless loops, like this:
> ```
> $ tree .
> .
> ├── dir1
> │   └── x -> ../dir2
> └── dir2
>     └── y -> ../dir1
> ```
With this option, this isn't endless since the target is inside the copied tree, a symlink is copied just as now.

**Comment by rssfed23** [comment]
> * Include the target of the symlinks in the backup

This is what I'm trying to do (backing up a bunch of servers that store DB metadata through a symlink off a SSD onto a slower HDD.
I'm relatively new to the awesomeness that is Restic, but is there any way for me to include the data inside the symlink within the same snapshot? 
I'm currently doing it in 2 different repos has has led to some inconsistencies when restoring (as I can't run the backups at the exact same time).

If anyone has a logical solution to the simple use case of following one symlink (that doesn't change or rotate or have the possibility of a loop) it would be much appreciated 

Aside from that, awesome work :)!

**Comment by 0x53A** [comment]
As this seems to be the main thread for symlink discussions, I'd like to add my use case:


I am using restic on Android through termux.

For convenience, termux creates a folder ``$HOME/storage``, which then contains ``dcim``, ``downloads``,``pictures``, etc.
All of these are symlinks.

I'd like to run restic against ``storage`` and tell it to just follow all symlinks.

As a workaround, I'm currently using ``--from-file`` with ``find ./storage/**/*`` (``from-file-raw`` does not exist, for some reason)

**Comment by rawtaz** [comment]
This issues seems to have diverged from "warn when backing up only symlinks" to "add ability for restic to follow symlinks".

**Comment by koalalorenzo** [comment]
UP! :D (hopefully this will be considered)

**Comment by nmaludy** [comment]
+1 would like to see this as well (--follow-symlinks or --dereference-symlinks)

**Comment by AngryMandragora** [comment]
I'm just in the process of migrating all Linux boxes from a custom Python backup script to restic (using runrestic), and I encountered a problem backing up VictoriaMetrics with it. VM creates a snapshot using soft links, but restic won't follow them, so I'm kinda stuck here. The data is 210GB and the only solution to use restic would be to copy it again to a local file system and back it up fom there (no sir, we don't have that much space on that server). Would restic follow the soft links I would have no problems at all (or by having a special read/write mount option, but I guess that's not possible).
TL;DR: please make this option available so I can keep my sanity :-)

**Comment by AngryMandragora** [comment]
> This issues seems to have diverged from "warn when backing up only symlinks" to "add ability for restic to follow symlinks".
Please see https://github.com/restic/restic/issues/2211#issuecomment-727549703, this is the place where we should discuss it.

**Comment by jhrcz-ls** [comment]
hi, i would like to have differrence for symlinks presented as source dirs and symlinks in the directory structure. there are cases (as somenone mentioned earlier) where you want to backup files referenced by symlinks somewhere in backed-up directory tree. i would usualy want to skip them because i symlink some specific files/dirs to different location and back them there. on the other hand i have biggest problem after relocating the top directory defined as source. adding slash at the end was intuitively something i expected to be the sollution and it was not, and rewrite of source definitions is required. what was realy annoing, is the fact, that the backup just runs (and realy fast ;o) and backup just the one symlink and no data. something you have to spod you get very bad surprise at some later moment.

**Comment by cccapon** [comment]
Would you consider having an option to follow symbolic links based on a pattern match (similar to the way exclude works)?  
Advantages:
- symbolic links can occur at any directory level, not just a top level folder
- include all symbolic links using '*'
- it is possible to exclude specific links and prevent infinite loop backups (a single option to include all symbolic links could not)
Disadvantages:
- it would be easy to make mistakes setting up your 'follow' pattern paths.

See: #3848

**Comment by mchangrh** [comment]
+1 for symlinks, I was using restic to back up a drive and when testing restores on the same machine, it appeared to restore all the files correctly but when I actually had to restore the data after data loss, a significant portion of the files were missing because the symlink destination was also dead.

Egg on my face for not testing on a separate identical machine and validating all the files but having symlink support or knowing about the lack thereof would have been nice

**Comment by bcspragu** [comment]
Just got bit by this (my own fault) with Kubernetes, which will mount Secrets + ConfigMaps as symlinks ([eg](https://stackoverflow.com/q/50685385)). I didn't notice until:

```bash
$ restic dump latest /config/Caddyfile
Fatal: cannot dump file: "/config/Caddyfile" should be a file, but is a "symlink"
```

Of course there are workarounds (`subPath` should work for me), but both a warning and a `--dereference` operation would be great here.

**Comment by ijprest-haptx** [comment]
Was bitten by this today.  +1 for a `--dereference` option.

**Comment by ijprest-haptx** [comment]
Also, semi-related:  I tried changing my symlinks from absolute paths to relative paths, so that the links were pointing to something *inside* my backup.  Unfortunately, Restic *always* restores relative `SYMLINKD` symlinks on Windows as `SYMLINK`... even if the original source was `SYMLINKD` or `JUNCTION`.  The result is unusable.
